packages/wordpress-playground-block/src/components/playground-preview/transpile-plugin-files.ts

```diff
@@ -0,0 +1,49 @@
+import type { EditorFile } from '../../index';
+
+let esbuild: Promise<typeof import('esbuild-wasm')> | undefined = undefined;
+let esbuildInitialized: any = undefined;
+
+export const transpilePluginFiles = async (
+	files: EditorFile[]
+): Promise<EditorFile[]> => {
+	if (esbuild === undefined) {
+		esbuild = import('esbuild-wasm');
+		esbuildInitialized = (await esbuild)!.initialize({
+			worker: true,
+			wasmURL: 'https://unpkg.com/esbuild-wasm@0.21.3/esbuild.wasm',
+		});
+	}
+
+	const transpiled = files.map(async (file) => {
+		// Don't transpile .js files
+		if (!file.name.endsWith('.js')) {
+			return file;
+		}
+		await esbuildInitialized;
+		try {
+			const transpiled = await (
+				await esbuild!
+			).transform(file.contents, {
+				loader: 'jsx',
+				target: 'esnext',
+				jsxFactory: 'wp.element.createElement',
+				format: 'esm',
+			});
+			return [
+				{
+					name: file.name + '.src',
+					contents: file.contents,
+				},
+				{
+					name: file.name,
+					contents: transpiled.code,
+				},
+			];
+		} catch {
+			// Default to an untranspiled file
+			return [file];
+		}
+	});
+	// Flatten the array
+	return (await Promise.all(transpiled)).flatMap((x) => x);
```

Comment:
I understand you may not be done with this, but...

If there are transpilation errors, can those be relayed to the user somehow?

Reply:
Good call! I just did this:

![CleanShot 2024-05-22 at 19 25 43@2x](https://github.com/WordPress/playground-tools/assets/205419/b2859f78-d9f3-4377-94b2-73f1b87e1143)

Reply:
That looks great!